### PR TITLE
Np 44561 create endpoint for updating ticket

### PR DESCRIPTION
**Pull request by EyobSikt**
Creating endpoint (PUT) for updating ticket assignee.

### CODE REVIEW

File: tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        if (ticketAssignee.getUser().equals(assignee)) {
+            findableDoiTicketSideEffects(requestInfo);
+        }
+    }
+
+    private void findableDoiTicketSideEffects(RequestInfo requestInfo)
+        throws NotFoundException, BadMethodException, BadGatewayException {
+        var publication = getPublication(requestInfo);
+        publicationSatisfiesDoiRequirements(publication);
+        createFindableDoiAndPersistDoiOnPublication(publication);
```
Comment (anetteOlli) [inline tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java:106]: Shouldn't the ticket get status "complete" if the publication has a findable DOI?

File: tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        throws UnauthorizedException {
+        return requestInfo.getCurrentCustomer().equals(ticket.getCustomerId());
+    }
+
+    private static boolean isAuthorizedToClaimTickets(RequestInfo requestInfo) {
+        return requestInfo.userIsAuthorized(AccessRight.APPROVE_DOI_REQUEST.toString());
+    }
+
+    private void doiTicketSideEffects(TicketDto input, final RequestInfo requestInfo)
+        throws NotFoundException, BadMethodException, BadGatewayException {
```
Comment (anetteOlli) [inline tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java:93]: From the open-api spec it seems to me that you submit a ticket with an assigne to update the database with. But from the code the assigne is fetched from the request info?

If the correct version is from request info, the open-api spec should not receive a body.

And if the body should be used so you can assign whoever you want, the findableDoiTicketSideEffects is a bit sketchy.

**Review by anetteOlli — CHANGES_REQUESTED** [review]

File: docs/openapi.yaml [hunk 3]
```diff
@@ -505,6 +505,77 @@ paths:
...
+          type: string
+        required: true
+    put:
+      description: Update the assignee of an existing ticket
+      requestBody:
+        description: A Ticket Object containing the new ticket assignee
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/Ticket'
```
Comment (anetteOlli) [inline docs/openapi.yaml:527]: schema: {}

File: tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    protected Void processInput(TicketDto input, RequestInfo requestInfo, Context context) throws ApiGatewayException {
+        var ticketIdentifier = extractTicketIdentifierFromPath(requestInfo);
+        var ticket = ticketService.fetchTicketByIdentifier(ticketIdentifier);
+
+        if (userIsNotAuthorized(requestInfo, ticket)) {
+            throw new ForbiddenException();
+        }
+        if (ticket instanceof DoiRequest) {
+            doiTicketSideEffects(input, requestInfo);
+        }
```
Comment (truhacevkir) [inline tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java:68]: Not sure if you have to include doi side effects here. When updating status for pubishingRequest, we have side effects to make doi findable/delete it. In your case you are adding assignee, that should not do anything with doi (as I see it)

File: tickets/src/test/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        var addAssigneeTicket = ticket.updateAssignee(publication, user.getUser());
+        var request = authorizedUserAssigneesTicket(addAssigneeTicket);
+        handler.handleRequest(request, output, CONTEXT);
+        var response = GatewayResponse.fromOutputStream(output, Void.class);
+        assertThat(response.getStatusCode(), is(equalTo(HTTP_ACCEPTED)));
+        var actualTicket = ticketService.fetchTicket(ticket);
+        assertThat(actualTicket.getAssignee(), is(equalTo(addAssigneeTicket.getAssignee())));
+        var publicationInDatabase = resourceService.getPublicationByIdentifier(publication.getIdentifier());
+        assertThat(publicationInDatabase.getDoi(), is(equalTo(publication.getDoi())));
+    }
```
Comment (truhacevkir) [inline tickets/src/test/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandlerTest.java:75]: Not sure if you have to test this, as your handler has nothing to do with DOI

**Review by truhacevkir — CHANGES_REQUESTED** [review]

**Comment by truhacevkir** [comment]
Do we want to update ticket assignee for COMPLETED ticket?

File: tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java
Comment (truhacevkir) [inline tickets/src/main/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandler.java:1]: Class name has to start with capital letter

File: tickets/src/test/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        return createAssigneeTicketHttpRequest(ticket, AccessRight.APPROVE_DOI_REQUEST, ticket.getCustomerId());
+    }
+
+    private InputStream createAssigneeTicketHttpRequest(TicketEntry ticket,
+                                                        AccessRight accessRight,
+                                                        URI customer) throws JsonProcessingException {
+        return new HandlerRequestBuilder<TicketDto>(JsonUtils.dtoObjectMapper)
+            .withBody(TicketDto.fromTicket(ticket))
+            .withAccessRights(customer, accessRight.toString())
+            .withCustomerId(customer)
```
Comment (truhacevkir) [inline tickets/src/test/java/no/unit/nva/publication/ticket/update/updateTicketAssigneeHandlerTest.java:150]: .withCustomerId() is deprecated, you can use withCurrentCustomer() instead

File: docs/openapi.yaml (hunk 3 above)
Comment (anetteOlli) [reply docs/openapi.yaml]: This comment is no longer valid.

**Pull request closed without merging**